pulsar-function-go/pf/context.go

```diff
@@ -119,6 +120,12 @@ func (c *FunctionContext) GetUserConfMap() map[string]interface{} {
 	return c.userConfigs
 }
 
+// NewOutputMessage send message to the topic
+// @param topicName: The name of the topic for output message
+func (c *FunctionContext) NewOutputMessage(topicName string) pulsar.Producer {
```

Comment:
why a `pulsar.Producer` is returned for this method?

in the corresponding java side, `public <O> TypedMessageBuilder<O> newOutputMessage(String topicName, Schema<O> schema)`

Reply:
This is because Java client and Go client have different packaging for producer and message. In the Go client, they are independent structs; in the Java client, they are unified encapsulated into TypedMessageBuilder objects.

But in fact, newOuputMessage essentially intercepts the link that sends messages to `outputTopic` during pulsar function processing. So we need to expose the producer to allow users to send messages out.

Reply:
I can understand the desire for naming parity, but I wonder whether or not it is better to call this something different, given how different the concepts are.  In Go, if I see `New<Foo>` I expect that I am going to be receiving a pointer to a new struct of that type, or at least an interface of some kind with that name.  In this case, I would expect to get a new Message.

I think I would probably recommend something like the following:

```go
func (c *FunctionContext) SendTopic(topicName string, message *pulsar.ProducerMessage)
```

or even
```go
func (c *FunctionContext) SendTopicMessage(topicName string, message *pulsar.ProducerMessage)
func (c *FunctionContext) SendTopic(topicName string, payload []byte)
```

which, given the way producers are currently packaged, might benefit from a registry of producers per topic:

```go
var producerRegistry = map[string]pulsar.Producer{}
func (gi *goInstance) getProducer(topicName string) (pulsar.Producer, error) {
        if found := producerRegistry[topicName]; found != nil {
                return found, nil
        }
	properties := getProperties(getDefaultSubscriptionName(
		gi.context.instanceConf.funcDetails.Tenant,
		gi.context.instanceConf.funcDetails.Namespace,
		gi.context.instanceConf.funcDetails.Name), gi.context.instanceConf.instanceID)


	producer, err := gi.client.CreateProducer(pulsar.ProducerOptions{
		Topic:                   topicName,
		Properties:              properties,
		CompressionType:         pulsar.LZ4,
		BatchingMaxPublishDelay: time.Millisecond * 10,
		// Set send timeout to be infinity to prevent potential deadlock with consumer
		// that might happen when consumer is blocked due to unacked messages
	})
	if err != nil {
		gi.stats.incrTotalSysExceptions(err)
		log.Errorf("create producer error:%s", err.Error())
		return nil, err
	}
        producerRegistry[topicName] = producer
	return producer, err
}
```

Reply:
Admittedly, there are certainly more idiomatic ways to handle some of these concepts. I am generally against using global variables, especially maps with no synchronization, but from what I can tell from the code all access to this map would be synchronous, so no problems there, and more profound changes to the Go functions SDK would require a PIP, which I'm certainly open to submitting down the road.

Reply:
Thanks @flowchartsman feedback. In fact, here we hope to expose the *Producer* interface rather than the `producerMessage` struct. In this way, users can use this interface to maximize their own behavior, including setting `producerOptions` when creating a producer. 

Again, `NewOutputMessage()` most fundamental purpose is to intercept the results processed by the Pulsar Function, and then transfer the processing power of this result to the user for customization.